Rebuilt long beer example. (#410)
w3c/sdw-sosa-ssn: ssn/rdf/examples/ACME-Beer-Example.md

* [Recording data using observation collection](#recording-data-using-observation-collection)
* [Recording data using sample collection](#recording-data-using-sample-collection)

The example is of interest because it involves the repurposing of a
commodity, a consumer-grade [sosa:Platform](https://www.w3.org/TR/vocab-ssn/#SOSAPlatform) deployed to
address a domain-specific problem, the movement of a platform from
environment to environment and the possible ambiguity of data-ownership in
each of these environments.  It also demonstrates the location ambiguity
that can sometimes be present when modeling sensor readings.

Note: The use of long identifiers is for readability purposes and is non-normative.

### A short note about sensors as commercial products

Sensors are often sourced
commercially and installed
as part of larger
[deployments]((https://www.w3.org/TR/vocab-ssn/#SOSADeployment) or
[platforms](https://www.w3.org/TR/vocab-ssn/#SOSAPlatform). These
standardized offerings need to be communicated in a way that is machine
readable and other vocabularies are available to do so.

> From a procurement and product recall approach, vocabulatories are

Reviewer: ```suggestion
From a procurement and product recall approach, vocabularies
```

differentiate between the product class and the specific instance of the
product being consumed.  The ability to align Sosa and product vocabularies
are beneficial in that it permits a unified description of the sensor.  This

> include modeling the sensor throughout its lifetime, announcing commercial

Reviewer: ```suggestion
includes modeling the sensor throughout its lifetime, announcing commercial
```

availability, procurement, deployment, operation, decomissioning and
disposal.

> To this end this example models IBS TH2 sensors using schema.org Product and

Reviewer: ```suggestion
To this end this example models IBS TH2 sensors using [Schema.org](https://schema.org/) Product and
```

Reviewer 2: Any reason not to link directly?
```suggestion
To this end this example models IBS TH2 sensors using [Schema.org Product](https://schema.org/Product) and
```

[GS1 WebVoc](https://ref.gs1.org/voc/) Product classes.  This permits concurrent support for both Search Engine
Optimization (SOA) and consumer supply chain product identification using
barcodes.

> Different design decisions also mean that design desision such as [punning](https://www.w3.org/2007/OWL/wiki/Punning)

Reviewer: ```suggestion
Different design decisions also mean that design decisions such as [punning](https://www.w3.org/2007/OWL/wiki/Punning)
```

GS1 deals with this same problem by subclassing its [Product](https://ref.gs1.org/voc/Product)
class for both unique instances or indicating lot membership. Concurrently
supporting both vocabularies obviously requires that a physical instance of
a sensor be both an instance and a class. 

## Packaging and initial storage
<img src="../../images/InkBird_IBS_TH2-packaging.png" width="25%" alt="The sensor platform and the beer are packaged into a container as a product">

Alice works for Acme Brewery Co.  She procures an InkBird IBS-TH2 Platform
(Serial Number: 12345) and places it inside a beer carton with six bottles
(a "six-pack") of Acme Brewery's famous Porter beers with a product lot code
of PO202402.  She seals the package and places it inside of Acme Brewery's
cooler to await shipment.

We model the packaging activity and the placement of the Platform into the
crate before it is sealed.  The platform monitors the ambient air
temperature and the relative humidity within the beer crate every few
minutes.  In the following snippet, we instantiate the IBS-TH2 Platform, the
carton, and the activity that represents the packaging of the beer and the

> creation of the beer carton as a product:

Reviewer: Anything missing after the colon?

    sosa:resultTime "2024-02-20T01:35:45Z"^^xsd:dateTime ;
    sosa:hasResult [
      a qudt:QuantityValue ;
      qudt:hasUnit unit:DEG_C ;
      qudt:value 12.0 ;
    ] ;
.
<1b/observation> rdf:type sosa:Observation ;
    rel:own <acmeBreweryCo> ;
    sosa:observedProperty sosa-env:AmbientHumidity;
    sosa:hasUltimateFeatureOfInterest <0001/ProductPackaging>;
    sosa:madeBySensor <12345/HumiditySensor>  ;
    sosa:hasFeatureOfInterest <10/PO202402/21/0001/acmePorterSixPackAirSample> ;
    sosa:resultTime "2024-02-20T01:35:45Z"^^xsd:dateTime ;
    sosa:hasResult [
      a qudt:QuantityValue ;
      qudt:hasUnit unit:PERCENT ;
      qudt:value 60 ;
    ] ;
.
<breweryObserver> a prov:Activity ;
    rdfs:comment "Brewery operating system logging process." ;
    prov:atLocation <acmeBreweryCo> ;
    prov:generated <1a/observation> ;
    prov:generated <1b/observation> ;
    prov:wasStartedBy <alice> . # She turned it on last time.     
```

In the example above, concurrent use of the 
[sosa:hasFeatureOfInterest](https://www.w3.org/TR/vocab-ssn/#SOSAhasFeatureOfInterest) 
and [sosa:hasUltimateFeatureOfInterest](https://www.w3.org/TR/vocab-ssn/#SOSAhasUltimateFeatureOfInterest) 

> properties is made to account for the repurposing of a generic sensor.  The

Reviewer: ```suggestion
object properties is made to account for the repurposing of a generic sensor.  The
```

actual measurement performed by the platform is the air temperature within
the carton, as a proxy for the beer temperature within each beer vessel. 
Modeling the full thermodynamic activity relating one measurement to the
other is beyond the scope of this document, but could be implemented as a
second order "virtual" [sosa:Sensor](https://www.w3.org/TR/vocab-ssn/#SOSASensor) or a sophisticated [sosa:Procedure](https://www.w3.org/TR/vocab-ssn/#SOSAProcedure).  Thus the use of the [sosa:hasUltimateFeatureOfInterest](https://www.w3.org/TR/vocab-ssn/#SOSAhasUltimateFeatureOfInterest) property allows for flexibility in interpretation of the measurement within a context *not necessarily intended by the original sensor design*.
   

> Physically, both <12345/HumiditySensor> and <12345/TemperatureSensor> are monitoring the ambient air within the specific Porter carton which we define as <10/PO202402/21/0001/acmePorterSixPackAirSample> which resolves the issue of the location of the measurement when the carton itself is in motion. The <a href="#SOSAhasUltimateFeatureOfInterest">`sosa:hasUltimateFeatureOfInterest`</a> property targets the beer temperature within that carton and the humidity the cardboard of the carton was subjected to:

Reviewer: ```suggestion
Physically, both <12345/HumiditySensor> and <12345/TemperatureSensor> are monitoring the ambient air within the specific Porter carton which we define as <10/PO202402/21/0001/acmePorterSixPackAirSample> which resolves the issue of the location of the measurement when the carton itself is in motion. The <a href="#SOSAhasUltimateFeatureOfInterest">`sosa:hasUltimateFeatureOfInterest`</a> object property targets the beer temperature within that carton and the humidity the cardboard of the carton was subjected to:
```

<10/PO202402/21/0001/acmePorterSixPackBeerSample> a  sosa:FeatureOfInterest, sosa:Sample, beer:Porter ;
    rdfs:label "A (virtual) sample of beer within the six pack"@en ;
    sosa:isSampleOf <10/PO202402/21/0001/acmePorterSixPack> .

# We define this as the beer temperature, but ontologically it is a 
# generic definition for the temperature of the sample.
<10/PO202402/21/0001/BeerTemperature> a sosa:Property ;
    qudt:hasUnit unit:DEG_C ;
    sosa:isPropertyOf <10/PO202402/21/0001/acmePorterSixPackBeerSample> ;
    rdfs:label "Biertemperatur"@de, "Beer Temperature"@en, "Température de la bière"@fr ;
    skos:definition "Temperature of Beer."@en .

```

The SOSA ontology takes a "feature and property" approach that is referenced
by observations.  The feature of interest &mdash; in this case, the ambient
air within the carton (<10/PO202402/21/0001/acmePorterSixPackAirSample>) is
the feature that the Platform is concerned with.  However, it is really a
proxy for both the temperature of the beer within the individual containers
(<10/PO202402/21/0001/acmePorterSixPackBeerSample>) within the carton and
the relative humidity to which the packaging (<0001/ProductPackaging>) is

> being exposed.  The beer "sample" in this case is virtual; the node serves

Reviewer: In my view this contradicts the definition of `sosa:Sample`.

Reviewer 3: I see no problem with notional samples as part of the workflow.

only to semantically link the observation to the beer liquid itself without
any physical sample being taken.  A more detailed example with physical beer
samples is found in [[[#beerCollections]]].

## Shipping to Retail Store<
<img src="../../images/InkBird_IBS_TH2-shipping.png" width="25%" alt="The beer carton is moved from the brewery to the retail location.">

After being chilled in the brewery cooler, the beer carton is then loaded on
one of the Acme delivery trucks for shipment to the retail store.  The

> truck's cooling management system is recording the sensor readings for both

Reviewer: ```suggestion
cooling management system of the truck is recording the sensor readings for both
```

the shipping company and the brewery.  An on-board GPS unit annotates the
activity of recording observations as the truck moves from location to
location.  This position information represents the location of the shipping
truck itself without reference to the beer carton or the platform generating
the readings.  This is important, in that the underlying properties and
features referenced by the sensors of the platform have not changed, even

> through the beer carton and the platform have obviously changed physical

Reviewer: ```suggestion
though the beer carton and the platform have obviously changed physical
```

    sosa:hasFeatureOfInterest <10/PO202402/21/0001/acmePorterSixPackAirSample> ;
    sosa:resultTime "2024-02-22T04:15:05Z"^^xsd:dateTime ;
    sosa:hasResult [
      a qudt:QuantityValue ;
      qudt:hasUnit unit:PERCENT ;
      qudt:value 74 ;
    ] ;
.
<123/TruckObserver> a prov:Activity ;
    rdfs:comment "Truck onboard monitoring system" ;
    prov:used <acmeTruck> ;
    prov:generated <67a/observation> ;
    prov:generated <67b/observation> ;
    prov:startedAtTime "2024-02-22T04:15:05Z"^^xsd:dateTime ;
    prov:endedAtTime "2024-02-22T05:55:38Z"^^xsd:dateTime ;
    prov:wasStartedBy <driver> ;
    prov:atLocation <gpsLocation> .

# This is the output from the truck's GPS receiver. We only get coordinates
# from it.

<gpsLocation> a prov:Location, geosparql:Geometry; 
   geosparql:asWKT "POINT(-79.35553 43.66372)"^^geosparql:wktLiteral .
```
 
The sensors of the platform are reporting values for the same properties of
the same features of interest; the physical location of the carton of beer
does not effect the process because the platform monitors the inside of the
carton itself.  Modeling its location and the process of (un)loading of the

> carton from the truck is done through other rdf nodes.  The truck on-board

Reviewer: ```suggestion
carton from the truck is done through other RDF nodes.  The truck on-board
```

monitoring system does report a GPS geometry as a Location.  This is the
location at which the data was recorded from the platform.  It is

> conceivable that the geometry node is shared with a vehicle tracking system

Reviewer: ```suggestion
conceivable that the geometry node is shared with a vehicle tracking system.
```

> rdf representation or the vehicle's delivery scheduling application but it

Reviewer: ```suggestion
RDF representation or the vehicle's delivery scheduling application but it
```

is not mandated by Sosa.

Note: Locations of Platform, Sensors, and measured samples are often
conflated in non-semantically enabled systems and the semantics often
implicitly assumed by the application.  The deep semantic modeling within
Sosa makes no such implicit assumptions and locations can be assigned to all
elements independently.
 
## Display in Retail Store Cooler 

<img src="../../images/InkBird_IBS_TH2-receiving.png" width="25%" alt="The beer carton is stored in the display cooler for retail sale.">

> When received from the delivery company by the retail store, the beer is displayed for sale in retail coolers which log the sensors to the data being held by the retail store. We notice again the semantics of location and containment; the sensor is contained within the carton, which is contained within the cooler, which is itself located within the store. However, none of these semantic is recorded. The only location activity is the activity is taking place within the supermarket cooler; not the supermarket itself. 

Reviewer: ```suggestion
When received from the delivery company by the retail store, the beer is displayed for sale in retail coolers which log the sensors to the data being held by the retail store. We notice again the semantics of location and containment; the sensor is contained within the carton, which is contained within the cooler, which is itself located within the store. However, none of these semantics are recorded. The only location activity is taking place within the supermarket cooler; not the supermarket itself. 
```

    sosa:madeBySensor <12345/TemperatureSensor>  ;
    sosa:hasFeatureOfInterest <10/PO202402/21/0001/acmePorterSixPackAirSample> ;
    sosa:resultTime "2024-02-22T06:00:13Z"^^xsd:dateTime ;
    sosa:hasResult [
      a qudt:QuantityValue ;
      qudt:hasUnit unit:DEG_C ;
      qudt:value 12 ;
    ] ;
.
<98b/observation> rdf:type sosa:Observation,prov:Entity ;
    rel:own <acmeSupermarket> ;
    sosa:observedProperty ex:airRelativeHumidity ;
    sosa:hasUltimateFeatureOfInterest <0001/ProductPackaging> ;
    sosa:madeBySensor <12345/HumiditySensor>  ;
    sosa:hasFeatureOfInterest <10/PO202402/21/0001/acmePorterSixPackAirSample> ;
    sosa:resultTime "2024-02-22T06:00:13Z"^^xsd:dateTime ;
    sosa:hasResult [
      a qudt:QuantityValue ;
      qudt:hasUnit unit:PERCENT ;
      qudt:value 55 ;
    ] ;
.
<CoolerMonitoringProcess> a prov:Activity ;
    rdfs:comment "Smart cooler monitoring system" ;
    prov:startedAtTime "2024-02-22T06:00:13Z"^^xsd:dateTime ;
    prov:atLocation <acmeSupermarketCooler> ;
    prov:generated <98a/observation> ;
    prov:generated <98b/observation> ;
    prov:wasStartedBy <acmeSupermarket> .

```

## Notes on deployment and observation aggregation
 

> In previous examples, <a href="https://www.w3.org/TR/prov-o/Activity">prov:Activity</a> was used as a means of aggregating results in an archiving reception. Other mechanisms exists to define the purpose of the System and to aggregate collections of results within specific contexts. 

Reviewer: ```suggestion
In previous examples, <a href="https://www.w3.org/TR/prov-o/Activity">prov:Activity</a> was used as a means of aggregating results in an archiving reception. Other mechanisms exist to define the purpose of the System and to aggregate collections of results within specific contexts. 
```
